mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala

```diff
@@ -341,6 +443,15 @@ class Completions(
               val include =
                 !isUninterestingSymbol(sym) &&
                   isNotLocalForwardReference(sym)
+                  && (!sym.info.typeSymbol.is(
+                    Flags.Module
+                  ) || (!isTypePosition && !isNewPosition))
+                  && (!isTypePosition || !sym.is(Flags.Method))
+                  // !sym.info.typeSymbol.is(Flags.Method) does not detect Java methods
+                  && (!isInstantiationOrMethodCallPos || (!sym.isClass && !sym.toString
+                    .startsWith(
+                      "trait "
+                    )))
```

Comment:
can we move this filtering logic into `findSuffix` or somewhere related to suffix thing?
I think `filterInteresting` should filter symbols purely by those symbols properties, and suffix specific logic shouldn't be here.

Reply:
well, indeed the `isInstantiationOrMethodCallPos` `isNewPosition` and `isTypePosition` have double use:

- for knowing whether to add the suffix or not,
- for knowing whether to keep a result or filter it out, for example, in the case that you've highlighted, we should not allow this code, regardless of what the suffix is, because `A` is a `trait` and it would be semantically invalid to put a `trait` in instantiation or method call positions.

```
trait A {}

object B{
  @@
}

```

as another case, we should not be putting `object`s in the `TypePosition`s. So for example, the following code is semantically invalid regardless of the suffix, because only `class` or `trait` can be types.

```
object A {
   def apply(): A = ???
 }

object B{
    val a: A@@ = null
}

```
Similarly, the apply method of `A` is a wrong completion for the type position.

Reply:
Ah, I realized this PR is doing filter out symbols based on positions, on top of the type-completion :+1:

Can you extract these lines into a method whose name is something we can tell it filters out invalid trait ... etc completions according to position?

Reply:
Precisely. I was actually doing it in the new case class you suggested, which was inspired by putting the suffix related methods there 👍